mono/mini/mini.h

```diff
@@ -3016,4 +3016,69 @@ mono_arch_load_function (MonoJitICallId jit_icall_id);
 MonoGenericContext
 mono_get_generic_context_from_stack_frame (MonoJitInfo *ji, gpointer generic_info);
 
+#if defined(__APPLE__) && defined(__arm64__)
+
+typedef enum {
+	JPM_NONE,
+	JPM_ENABLED,
+	JPM_DISABLED,
+} jit_protect_mode;
+
+extern __thread jit_protect_mode arm_current_jit_protect_mode;
+
+static void mono_arm_jit_write_protect_enable()
+{
+	if (__builtin_available(macOS 11, *)) {
+		if (arm_current_jit_protect_mode != JPM_ENABLED) {
+			pthread_jit_write_protect_np(1);
+			arm_current_jit_protect_mode = JPM_ENABLED;
+		}
+	}
+}
+
+static void mono_arm_jit_write_protect_disable()
+{
+	if (__builtin_available(macOS 11, *)) {
+		if (arm_current_jit_protect_mode != JPM_DISABLED) {
+			pthread_jit_write_protect_np(0);
+			arm_current_jit_protect_mode = JPM_DISABLED;
+		}
+	}
+}
+
+#define MONO_SCOPE_ENABLE_JIT_WRITE()					\
+	__attribute__((unused, cleanup(mono_arm_restore_jit_protect_mode))) \
+	jit_protect_mode scope_restrict_mode = arm_current_jit_protect_mode; \
+        mono_arm_jit_write_protect_disable();					\
```

Comment:
We don't really use `__attribute__((cleanup(...)))` in Mono, so hiding it in this macro is unexpected and may lead to surprises when reasoning about the thread state.  But on the other hand it makes the patch quite compact.

Reply:
Like a local in C++ with destructor.